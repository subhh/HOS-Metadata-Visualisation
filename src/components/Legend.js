import React, { useContext, useCallback } from 'react'
import { Route } from 'react-router-dom'
import styled from 'styled-components'
import { Box as ReBox, Text, Heading } from 'rebass'
import { StateContext, DispatchContext } from 'state'
import { transparentize } from 'polished'
import {
  toggleMapLegend,
  toggleGraphLegend,
  toggleGraphExploreLegend,
} from 'state/actions'
import {
  fill,
  scaledWidth,
  scaledHeight,
  centerSVG,
  invert,
  shadow,
} from 'utils'
import { ReactComponent as Help } from 'icons/help.svg'
import { ReactComponent as Collapse } from 'icons/expand.svg'
import { ReactComponent as LegendSVG } from 'icons/legend.svg'
import { ReactComponent as LegendMap } from 'icons/legend-map.svg'

const Container = styled(ReBox).attrs(({ theme, collapsed }) => ({
  fontSize: 1,
  bg: collapsed ? 'transparent' : transparentize(0.05, theme.colors.white),
  ml: 4,
  mb: 5,
}))`
  /* transition: border-radius 100ms ease-in-out; */
  /* border-radius: ${({ collapsed }) => (collapsed ? 48 : 0)}px; */
  position: absolute;
  z-index: 1;
  bottom: 0;
`

const IconWrapper = styled(ReBox).attrs({
  p: 12,
  width: 40,
  height: 40,
})`
  transition: all 0.2s ease-in-out;
  ${centerSVG};
  ${scaledWidth};
  ${scaledHeight};
`

const InlineIcon = styled(IconWrapper)`
  position: absolute;
  right: 0;
  top: 0;

  :hover svg {
    fill: ${({ theme }) => theme.colors.blueMediumLight};
  }
`

const StandAloneIcon = styled(IconWrapper).attrs({
  bg: 'ghostWhite',
})`
  border-radius: 40px;
  ${shadow};

  :hover {
    ${invert};
  }
`

const CloseIcon = styled(Collapse)`
  ${fill};
`

const HelpIcon = styled(Help)`
  ${fill};
`

const LegendHeading = ({ children }) => (
  <Heading fontSize={3} color="blueDark" mb={4}>
    {children}
  </Heading>
)

const LegendText = props => (
  <Text lineHeight={1.5} color="greyDark" width={330} {...props} />
)

const ContentToggle = ({ path, show, actionCreator, title, children }) => {
  const dispatch = useContext(DispatchContext)
  const handleClick = useCallback(() => dispatch(actionCreator()), [
    actionCreator,
    dispatch,
  ])

  return (
    <Route
      exact
      path={path}
      render={() => (
        <Container collapsed={!show} p={show ? 4 : 0}>
          {show ? (
            <InlineIcon data-testid="toggle" onClick={handleClick}>
              <CloseIcon fill="greyDark" />
            </InlineIcon>
          ) : (
            <StandAloneIcon data-testid="toggle" onClick={handleClick}>
              <HelpIcon fill="blueMediumDark" />
            </StandAloneIcon>
          )}

          {show && (
            <>
              <LegendHeading>{title}</LegendHeading>
              {children}
            </>
          )}
        </Container>
      )}
    />
  )
}

/**
 * Floating container that contains introductory remarks and descriptive information/**
 */
const Legend = ({ width, height }) => {
  const { showMapLegend, showGraphLegend, showGraphExploreLegend } = useContext(
    StateContext
  )

  return (
    <>
      <ContentToggle
        path="/map"
        show={showMapLegend}
        actionCreator={toggleMapLegend}
        title="Einrichtungen"
      >
        <LegendText mb={4}>
          Einrichtungen stellen Hamburg Open Science Daten über ihre
          wissenschaftlichen Publikationen und Forschungsdaten zur Verfügung.
          Die Karte zeigt die Standorte und die Anzahl der zum Suchbegriff
          gefundenen Publikationen.
        </LegendText>
        <LegendMap />
      </ContentToggle>
      <ContentToggle
        path="/graph"
        show={showGraphLegend}
        actionCreator={toggleGraphLegend}
        title="Kollektionen"
      >
        <LegendText>
          Kollektionen sind Daten-Sammlungen von Publikationen. Mehrere
          Kollektionen können zu einer Einrichtung gehören. Hamburg Open Science
          werden Daten aus verschiedenen Speicher-Systemen zur Verfügung
          gestellt. Dargestellt wird die Anzahl der in den Kollektionen
          verzeichneten Publikationen.
        </LegendText>
      </ContentToggle>
      <ContentToggle
        path="/graph/explore"
        show={showGraphExploreLegend}
        actionCreator={toggleGraphExploreLegend}
        title="Legende"
      >
        <LegendSVG />
      </ContentToggle>
    </>
  )
}

export default Legend
