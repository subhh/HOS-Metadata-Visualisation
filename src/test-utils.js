import React, { useReducer } from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { MemoryRouter } from 'react-router-dom'
import { StateContext, DispatchContext, reducer, initialState } from './state'
import theme from './theme'

const AllTheProviders = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)

  return (
    <ThemeProvider theme={theme}>
      <StateContext.Provider value={state}>
        <DispatchContext.Provider value={dispatch}>
          {children}
        </DispatchContext.Provider>
      </StateContext.Provider>
    </ThemeProvider>
  )
}

const WithRouter = ({ children }) => (
  <MemoryRouter>
    <AllTheProviders>{children}</AllTheProviders>
  </MemoryRouter>
)

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export const customRenderWithRouter = (ui, options) =>
  render(ui, {
    wrapper: WithRouter,
    ...options,
  })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }

export const mockAsync = data => {
  jest.spyOn(global, 'fetch').mockImplementation(url => {
    return Promise.resolve({
      json: () => Promise.resolve(data),
    })
  })
}

export const restoreAsyncMock = () => {
  global.fetch.mockRestore()
}
