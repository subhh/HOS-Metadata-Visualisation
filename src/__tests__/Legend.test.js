import React from 'react'
import Legend from 'components/Legend'
import { MemoryRouter } from 'react-router-dom'
import { render, fireEvent } from 'test-utils'

it('should match snapshot on route ‘/map’', () => {
  const { asFragment, getByTestId } = render(
    <MemoryRouter initialEntries={['/map']}>
      <Legend />
    </MemoryRouter>
  )

  expect(asFragment()).toMatchSnapshot()
  fireEvent.click(getByTestId('toggle'))
  expect(asFragment()).toMatchSnapshot()
})

it('should match snapshot on route ‘/graph’', () => {
  const { asFragment, getByTestId } = render(
    <MemoryRouter initialEntries={['/graph']}>
      <Legend />
    </MemoryRouter>
  )

  expect(asFragment()).toMatchSnapshot()
  fireEvent.click(getByTestId('toggle'))
  expect(asFragment()).toMatchSnapshot()
})

it('should match snapshot on route ‘/graph/explore’', () => {
  const { asFragment, getByTestId } = render(
    <MemoryRouter initialEntries={['/graph/explore']}>
      <Legend />
    </MemoryRouter>
  )

  expect(asFragment()).toMatchSnapshot()
  fireEvent.click(getByTestId('toggle'))
  expect(asFragment()).toMatchSnapshot()
})
