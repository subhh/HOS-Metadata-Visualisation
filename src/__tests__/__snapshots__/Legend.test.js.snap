// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot on route ‘/graph/explore’ 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa sc-htoDjs cKuKND"
    font-size="1"
  >
    <div
      class="sc-bdVaJa sc-dnqmqq sc-gZMcBi hvkIQp"
      data-testid="toggle"
      height="40"
      width="40"
    >
      <svg
        class="sc-VigVT kxnxSC"
        fill="blueMediumDark"
      >
        help.svg
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot on route ‘/graph/explore’ 2`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa sc-htoDjs coMsRC"
    font-size="1"
  >
    <div
      class="sc-bdVaJa sc-dnqmqq sc-iwsKbI lpfzsy"
      data-testid="toggle"
      height="40"
      width="40"
    >
      <svg
        class="sc-gqjmRU diPeGw"
        fill="greyDark"
      >
        expand.svg
      </svg>
    </div>
    <h2
      class="sc-bdVaJa sc-htpNat sc-bxivhb eydduH"
      color="blueDark"
      font-size="3"
      font-weight="bold"
    >
      Legende
    </h2>
    <svg>
      legend.svg
    </svg>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot on route ‘/graph’ 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa sc-htoDjs cKuKND"
    font-size="1"
  >
    <div
      class="sc-bdVaJa sc-dnqmqq sc-gZMcBi hvkIQp"
      data-testid="toggle"
      height="40"
      width="40"
    >
      <svg
        class="sc-VigVT kxnxSC"
        fill="blueMediumDark"
      >
        help.svg
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot on route ‘/graph’ 2`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa sc-htoDjs coMsRC"
    font-size="1"
  >
    <div
      class="sc-bdVaJa sc-dnqmqq sc-iwsKbI lpfzsy"
      data-testid="toggle"
      height="40"
      width="40"
    >
      <svg
        class="sc-gqjmRU diPeGw"
        fill="greyDark"
      >
        expand.svg
      </svg>
    </div>
    <h2
      class="sc-bdVaJa sc-htpNat sc-bxivhb eydduH"
      color="blueDark"
      font-size="3"
      font-weight="bold"
    >
      Kollektionen
    </h2>
    <div
      class="sc-bdVaJa sc-htpNat cvNkTe"
      color="greyDark"
      width="330"
    >
      Kollektionen sind Daten-Sammlungen von Publikationen. Mehrere Kollektionen können zu einer Einrichtung gehören. Hamburg Open Science werden Daten aus verschiedenen Speicher-Systemen zur Verfügung gestellt. Dargestellt wird die Anzahl der in den Kollektionen verzeichneten Publikationen.
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot on route ‘/map’ 1`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa sc-htoDjs cKuKND"
    font-size="1"
  >
    <div
      class="sc-bdVaJa sc-dnqmqq sc-gZMcBi hvkIQp"
      data-testid="toggle"
      height="40"
      width="40"
    >
      <svg
        class="sc-VigVT kxnxSC"
        fill="blueMediumDark"
      >
        help.svg
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot on route ‘/map’ 2`] = `
<DocumentFragment>
  <div
    class="sc-bdVaJa sc-htoDjs coMsRC"
    font-size="1"
  >
    <div
      class="sc-bdVaJa sc-dnqmqq sc-iwsKbI lpfzsy"
      data-testid="toggle"
      height="40"
      width="40"
    >
      <svg
        class="sc-gqjmRU diPeGw"
        fill="greyDark"
      >
        expand.svg
      </svg>
    </div>
    <h2
      class="sc-bdVaJa sc-htpNat sc-bxivhb eydduH"
      color="blueDark"
      font-size="3"
      font-weight="bold"
    >
      Einrichtungen
    </h2>
    <div
      class="sc-bdVaJa sc-htpNat elofON"
      color="greyDark"
      width="330"
    >
      Einrichtungen stellen Hamburg Open Science Daten über ihre wissenschaftlichen Publikationen und Forschungsdaten zur Verfügung. Die Karte zeigt die Standorte und die Anzahl der zum Suchbegriff gefundenen Publikationen.
    </div>
    <svg>
      legend-map.svg
    </svg>
  </div>
</DocumentFragment>
`;
